Sources/TestSpy/TestSpy/Predicates/CallstackMatcher.swift

```diff
@@ -6,99 +6,46 @@
 //  Copyright © 2018 Franco Meloni. All rights reserved.
 //
 
-public enum CallstackMatcher<T: Equatable>: CallstackPredicate {
+public enum CallstackMatcher<Method: Equatable> {
     case times(Int)
     case atLeast(times: Int)
     case never
     case any
     case only
-    case before(T)
-    case immediatelyBefore(T)
-    case after(T)
-    case immediatelyAfter(T)
-    
-    public func check(method: T, against callstack: [T]) -> Bool {
-        switch self {
-        case .times(let times):
-            return callstack.filter { $0 ==  method}.count == times
-        case .atLeast(let times):
-            return callstack.filter { $0 ==  method}.count >= times
-        case .never:
-            return !callstack.contains(method)
-        case .any:
-            return callstack.contains(method)
-        case .only:
-            return callstack == [method]
-        case .before(let otherMethod):
-            return check(method: method, isBefore: otherMethod, onCallstack: callstack)
-        case .immediatelyBefore(let otherMethod):
-            return check(method: method, isImmediatelyBefore: otherMethod, onCallstack: callstack)
-        case .after(let otherMethod):
-            return check(method: method, isAfter: otherMethod, onCallstack: callstack)
-        case .immediatelyAfter(let otherMethod):
-            return check(method: method, isImmediatelyAfter: otherMethod, onCallstack: callstack)
-        }
-    }
-    
-    public func description(forMethod method: T) -> String {
+    case before(Method)
+    case immediatelyBefore(Method)
+    case after(Method)
+    case immediatelyAfter(Method)
+
+    func predicate(method: Method) -> AnyCallstackPredicate<Method> {
```

Comment:
This might read better as 
```suggestion
    func predicate(for: Method) -> AnyCallstackPredicate<Method> {
```
This makes explicit that the passed method is the one you want to check against

Reply:
```suggestion
    func predicate(for method: Method) -> AnyCallstackPredicate<Method> {
```

That sound great, working on it.

Reply:
As 7ca79a2, rename into `predicate(for:)`